src/lib/plugins/PluginManager.ts

```diff
@@ -0,0 +1,64 @@
+import type { ClientOptions } from 'discord.js';
+import type { SapphireClient } from '../SapphireClient';
+import { PluginHook } from '../types/Enums';
+import type { Plugin } from './Plugin';
+import { postInitialization, postLogin, preInitialization, preLogin } from './symbols';
+
+export interface SapphirePluginHook {
+	(this: SapphireClient, options?: ClientOptions): unknown;
+}
+
+export interface SapphirePluginHookEntry {
+	hook: SapphirePluginHook;
+	type: PluginHook;
+	name?: string;
+}
+
+export class PluginManager {
+	public readonly registeredPlugins = new Set<SapphirePluginHookEntry>();
+
+	public registerPluginHook(hook: SapphirePluginHook, type: PluginHook, name?: string) {
+		if (typeof hook !== 'function') throw new TypeError(`The provided hook ${name ? `(${name}) ` : ''}is not a function`);
+		this.registeredPlugins.add({ hook, type, name });
+		return this;
+	}
+
+	public registerPreInitializationPluginHook(hook: SapphirePluginHook, name?: string) {
+		return this.registerPluginHook(hook, PluginHook.PreInitialization, name);
+	}
+
+	public registerPostInitializationPluginHook(hook: SapphirePluginHook, name?: string) {
+		return this.registerPluginHook(hook, PluginHook.PostInitialization, name);
+	}
+
+	public registerPreLoginPluginHook(hook: SapphirePluginHook, name?: string) {
+		return this.registerPluginHook(hook, PluginHook.PreLogin, name);
+	}
+
+	public registerPostLoginPluginHook(hook: SapphirePluginHook, name?: string) {
+		return this.registerPluginHook(hook, PluginHook.PostLogin, name);
+	}
+
+	public usePlugin(plugin: typeof Plugin) {
```

Comment:
plugins.usePlugin?

Reply:
They could call that ofc but most Klasa users are used to having .use on the client as well.
So I didn't want to break their norm more than I already did.

Reply:
This is not Client, this is PluginManager.

Reply:
Oh fair point.